lib/elixir/lib/module.ex

```diff
@@ -528,37 +526,73 @@ defmodule Module do
 
   # Simplify signatures to be stored in docs
 
-  defp simplify_signature({:\\, _, [left, right ]}, i) do
-    {:\\, [], [simplify_signature(left, i), right]}
+  defp simplify_signature(signature) do
+    folder = fn(el, acc) -> simplify_signature(el, acc) end
+
+    {_, {:collect, acc}} = :lists.mapfoldl(folder, {:collect, []}, signature)
+    {signature, _}       = :lists.mapfoldl(folder, {:generate, acc, acc}, signature)
+
+    signature
   end
 
-  defp simplify_signature({:%, _, [left, _]}, i) when is_atom(left) do
-    struct_name = camelcase_to_underscore(List.last(split(left)))
-    {:"#{struct_name}#{i}", [], nil}
+  defp simplify_signature({:\\, _, [left, right ]}, acc) do
+    {left, acc} = simplify_signature(left, acc)
+    {{:\\, [], [left, right]}, acc}
   end
 
-  defp simplify_signature({:=, _, [_, right]}, i) do
-    simplify_signature(right, i)
+  defp simplify_signature({:=, _, [_, right]}, acc) do
+    simplify_signature(right, acc)
   end
 
-  defp simplify_signature({var, _, atom}, _i) when is_atom(atom) do
+  defp simplify_signature({var, _, atom}, acc) when is_atom(atom) do
     case Atom.to_string(var) do
-      "_" <> rest -> {String.to_atom(rest), [], Elixir}
-      _           -> {var, [], nil}
+      "_" <> rest -> {{String.to_atom(rest), [], Elixir}, acc}
+      _           -> {{var, [], nil}, acc}
+    end
+  end
+
+  defp simplify_signature({:%, _, [left, _]}, acc) when is_atom(left) do
+    struct_name = camelcase_to_underscore(List.last(split(left)))
```

Comment:
The bug is coming from here. `camelcase_to_underscore` is returning a string which later on may end-up being used as key. We need to convert it to an atom before calling autogenerated.